### PR TITLE
feat(JSAM-42): implement database actions for login

### PR DESCRIPTION
**Pull request by earvinLi**
> **Finish JIRA Story: [JSAM-42](https://jira.greenfox.academy/browse/JSAM-42)**

### Finish:
- Add the word `production` to the end of `run deploy` script
- Delete environment variable declaring in `ecosystem.config.js` file
- Refactor login endpoint to extend it to use the migrated database so now we have a router plus data service workflow
- Write unit tests for login endpoint with mock data and mock function
- Change migrated data phone numbers all different

### To Do:
- Mock the MySQL connection

### CODE REVIEW

File: JSA-Mawsitsit-Backend/services/dataService.js [hunk 1]
```diff
@@ -8,6 +9,24 @@ const mysqlPromisedQuery = (dataConnection, queryStatement, queryInput = []) =>
...
+  const loginQueryStatement = `
+    SELECT password FROM users WHERE ${userIdentifier.includes('@') ? 'email' : 'phone_number'} = ?
+  `;
+  const loginQueryInput = [userIdentifier];
+
+  const loggedInUserPassword = await mysqlPromisedQuery(
+    mysqlConnection,
+    loginQueryStatement,
+    loginQueryInput,
+  ).catch((error) => error);
```
Comment (matecserven) [inline JSA-Mawsitsit-Backend/services/dataService.js]: what happens to this error? either send it to console or throw it again, if any problem occurs with the DB we know nothing about it

File: JSA-Mawsitsit-Backend/services/dataService.js [hunk 2]
```diff
@@ -26,4 +45,10 @@ const checkIdentifier = async (user) => {
...
 };
 
-module.exports = { registerQuery, checkIdentifier };
+module.exports = {
+  checkIdentifier,
+  loginUser,
+  registerQuery,
+};
+
+module.exports = { loginUser };
```
Comment (binjiezhang0626) [inline JSA-Mawsitsit-Backend/services/dataService.js]: duplicated export

File: JSA-Mawsitsit-Backend/services/__mocks__/dataService.js [hunk 3]
```diff
@@ -1,34 +1,51 @@
...
   },
 ];
 
+// Mock Functions
+const loginUser = async (userIdentifier, password) => {
+  const loggedInUser = mockUsers.find((user) => userIdentifier === user[
+    userIdentifier.includes('@') ? 'email' : 'phone_number'
+  ]);
+
+  if (!loggedInUser) return 'User doesn\' exit. Please check your username.';
```
Comment (matecserven) [inline JSA-Mawsitsit-Backend/services/__mocks__/dataService.js]: This is code duplication

File: JSA-Mawsitsit-Backend/services/dataService.js [hunk 4]
```diff
@@ -8,6 +9,24 @@ const mysqlPromisedQuery = (dataConnection, queryStatement, queryInput = []) =>
...
+  `;
+  const loginQueryInput = [userIdentifier];
+
+  const loggedInUserPassword = await mysqlPromisedQuery(
+    mysqlConnection,
+    loginQueryStatement,
+    loginQueryInput,
+  ).catch((error) => error);
+
+  if (loggedInUserPassword.length === 0) return 'User doesn\' exit. Please check your username.';
```
Comment (matecserven) [inline JSA-Mawsitsit-Backend/services/dataService.js]: put this logic in a function so you can reuse it in the test

File: JSA-Mawsitsit-Backend/routes/loginRouter.js [hunk 5]
```diff
@@ -1,30 +1,32 @@
...
   if (req.headers['content-type'] !== 'application/json') {
-    res.status(415).json({
-      message: 'Content-type must be application/json.',
-    });
-    return;
+    return res.status(415).send({ error: 'Content-type must be application/json.' });
   }
 
-  const checkMessage = checkLogin(userIdentifier, password);
+  if (!password) return res.status(400).send({ error: 'Please full in your password.' });
```
[the remark this answers is not in the document]
Comment (earvinLi) [reply JSA-Mawsitsit-Backend/routes/loginRouter.js]: this is too hard to fix noooooooooo 😭😭😭

File: JSA-Mawsitsit-Backend/services/validation.js [hunk 6]
```diff
@@ -3,21 +3,23 @@ const validateEmail = (email) => !email || email.includes('@');
 const validatePhoneNumber = (phoneNumber) => !phoneNumber || phoneNumber.length >= 8;
 const validatePassword = (password) => !password || weakRegex.test(password);
 
+const validateLogin = (inputPassword, userToLogin) => {
```
Comment (matecserven) [inline JSA-Mawsitsit-Backend/services/validation.js]: result = {
  success: '',
  error: ''
}

return result

File: JSA-Mawsitsit-Backend/routes/loginRouter.js [hunk 7]
```diff
@@ -1,30 +1,32 @@
...
-    res.status(200).json({
-      user_id: 'id',
-      checkMessage,
-    });
-  } else {
-    res.status(400).json({
-      checkMessage,
-    });
+  const loginUserMessage = await loginUser(userIdentifier, password);
+  if (loginUserMessage !== 'Welcome!!!') {
```
Comment (matecserven) [inline JSA-Mawsitsit-Backend/routes/loginRouter.js]: if querySucces == false {
send 500 + 'Internal server error'
return;
}
if result.error {
send 400 + user error message
}
if result.success {
send 200 + token
}

File: JSA-Mawsitsit-Backend/services/dataService.js [hunk 8]
```diff
@@ -8,6 +10,21 @@ const mysqlPromisedQuery = (dataConnection, queryStatement, queryInput = []) =>
...
+  const loginQueryStatement = `
+    SELECT * FROM users WHERE ${userIdentifier.includes('@') ? 'email' : 'phone_number'} = ?
+  `;
+  const loginQueryInput = [userIdentifier];
+
+  const userToLogin = await mysqlPromisedQuery(
+    mysqlConnection,
+    loginQueryStatement,
+    loginQueryInput,
+  ).catch((error) => error);
```
Comment (matecserven) [inline JSA-Mawsitsit-Backend/services/dataService.js]:
```suggestion
  ).catch((error) => {
  console.error(error);
  return false;
  }
```

File: JSA-Mawsitsit-Backend/services/dataService.js [hunk 9]
```diff
@@ -8,6 +10,21 @@ const mysqlPromisedQuery = (dataConnection, queryStatement, queryInput = []) =>
...
+
+  return validateLogin(inputPassword, userToLogin[0]);
```
Comment (matecserven) [inline JSA-Mawsitsit-Backend/services/dataService.js]: return userToLogin && validateLogin(inputPassword, userToLogin[0]);

Comment (earvinLi) [reply JSA-Mawsitsit-Backend/services/dataService.js]: A user types a wrong username will make this `undefined` not `false`, so we need to write `return Boolean(userToLogin) && validateLogin(inputPassword, userToLogin[0]);` with `Boolean` surrounding `userToLogin`  to make the suggestion above work, handling even more unpredictable USER NOT FOUND error. But we cannot know this `undefined` is from user typing a wrong username or there is something wrong with our server. Also I think the `catch` after `loginUser` already handle all possible server errors and we have a `Internal server error.` message.

**Pull request merged**